### ISSUE TITLE
Download option

### ISSUE THREAD

**Issue by yant0** [vfsfitvnm/ViMusic#83]
### Feature description

Ive read some issues on this particular feature such as #9 but i want this feature because i already have a collection of downloaded songs on my phone. A double database of songs that i listen to would be unconventional.

im sorry for bothering you about this again but please do reconsider.

### Why do you want this feature?

> Ive read some issues on this particular feature such as #9 but i want this feature because i already have a collection of downloaded songs on my phone. A double database of songs that i listen to would be unconventional.

### Additional information

m4a feature would be nice. Maybe even flac if there is any on yt music

**Comment by vfsfitvnm** [comment]
Hmm, do you want a download option or the possibility to play music files?
Check this source for audio files formats: https://gist.github.com/sidneys/7095afe4da4ae58694d128b1034e01e2

**Comment by yant0** [comment]
Both are a great feature to have but for this for this issue, i want to have the option to download tracks/albums. Other file formats are always appriciated but if its not possible/don't feel like implementing it, im satisfied with just the download feature.

**Comment by edzaa** [comment]
I think that option for downloading music in mp3 format would be great.

**Comment by vfsfitvnm** [comment]
Relevant: https://github.com/vfsfitvnm/ViMusic/issues/178#issuecomment-1207361949

**Comment by JFMugen** [comment]
@vfsfitvnm Is it possible to find/see cache location without Root? 

I am trying to whitelist cache location, but couldn't find it anywhere.

**Comment by vfsfitvnm** [comment]
The cache is in `/data/data/it.vfsfitvnm.vimusic/cache`. Why do you want to whitelist that path?

**Comment by JFMugen** [comment]
> The cache is in `/data/data/it.vfsfitvnm.vimusic/cache`. Why do you want to whitelist that path?

My phone has a low storage. I LOVE how the app automatically plays from cache without needing to download before or after, but since I can't move cache to SD card, once I clean the caches my saved songs goes away as well. 

Looks like this workaround works for me tho.

**Comment by vfsfitvnm** [comment]
The cache will **always** remain where it currently is. However I guess you should rethink of the way you clean the cache of your installed apps - I mean - this is not something that can be solved by a download feature or by the developer(s) of the app you installed, I'm afraid.

**Comment by JFMugen** [comment]
> The cache will **always** remain where it currently is. However I guess you should rethink of the way you clean the cache of your installed apps - I mean - this is not something that can be solved by a download feature or by the developer(s) of the app you installed, I'm afraid.

Yeah, I wasn't requesting a change/feature. I know it is something about my phone. I am using SDMaid for now to clean things time to time, and whitelisting worked for me. So no complaints.

(app is awesome btw, I love how buttersmooth it is :)

**Comment by JFMugen** [comment]
> Maybe some kind of second cache/database where only explicitly downloaded songs reside to have some differentiation build in would be an idea?

Recent update added an "Offline" tab that shows all cached songs for offline use. I think this is exactly what you want?

**Comment by Nesaijn** [comment]
You mean the "Offline"-Playlist? That is indeed helpful.
The differentiation between the cached songs and explicitly downloaded songs was just an implementation/functionality idea of mine for the download feature. I thought that it would make the management of the songs easier regarding the limit of cache max size, cleanup etc.

**Comment by uGeek** [comment]
How about adding the cache option inside a list. To download the list to cache
![Uploading IMG_20220812_172832.jpg…]()

**Comment by uGeek** [comment]
How about adding the cache option inside a list. To download the list to cache
![Uploading IMG_20220812_172832.jpg…]()

**Comment by uGeek** [comment]
![Uploading IMG_20220812_172832.jpg…]()

**Comment by uGeek** [comment]
![Uploading IMG_20220812_172832.jpg…]()

**Comment by uGeek** [comment]
![Uploading IMG_20220812_172832.jpg…]()

**Comment by JFMugen** [comment]
Please use Imgur or sth else

**Comment by uGeek** [comment]
https://tlgur.com/d/gwqJJQy8

Sorry, I'm having connection problems and now I realize how nice it would be to download cache lists. I was not able to upload an image

**Comment by vfsfitvnm** [comment]
(I deleted your duplicate comments)

I'm not incline to implement a classic "download feature", which is, download and export to the external storage. Instead, a "save to offline" or "pre cache" feature is more consistent.

**Comment by uGeek** [comment]
Yes, totally agree. that you can download your favorite lists by going to an area without internet, but that they are cached, only accessible from the application. Do not download an mp3 file in a directory. 

If you go on a trip, for example, you can download your lists. This way you do not depend on having internet.

 Is there now a way to delete songs or playlists from the cache?

**Comment by vfsfitvnm** [comment]
Yes, long press a song within the home page or the offline playlist, then select "Hide"

**Comment by uGeek** [comment]
> Yes, long press a song within the home page or the offline playlist, then select "Hide"

Ok Thanks. Great

**Comment by caveman1973** [comment]
Check this source for audio files formats: https://gist.github.com/sidneys/7095afe4da4ae58694d128b1034e01e2

There's 3 ranges of bitrate for WebM:
- 50k
- 70k
- 160k

m4a have:
- 48k
- 128k
- 256k

Since Vimusic uses the highest WebM quality. Isn't it possible to use m4a to enjoy AAC 256k?

**Comment by vfsfitvnm** [comment]
Not possible (#78)

**Comment by emkay5771** [comment]
Is it possible to add songs to the offline playlist manually, or to make a playlist available offline? Thanks

**Comment by arfanmrafeek** [comment]
Sorry, Found better alternative app for download option, Let VImusic be itself.

Next time someone ask this feature mention this project.

https://sangwan5688.github.io/
App name: Blackhole

**Comment by Claudi0-V** [comment]
> Sorry, Found better alternative app for download option, Let VImusic be itself.
> 
> Next time someone ask this feature mention this project.
> 
> https://sangwan5688.github.io/ App name: Blackhole

@arfanmrafeek Blackhole is stopping development (see BlackHole's [344](https://github.com/Sangwan5688/BlackHole/issues/344) ).

> I'm not incline to implement a classic "download feature", which is, download and export to the external storage. Instead, a > "save to offline" or "pre cache" feature is more consistent.

@vfsfitvnm first of all I really appreciate this app and the work that you've put on it, it is truly great. regarding the comment that i'm quoting, that would be great. an example of this use is my aunt, who works where there's no web access and would enjoy pre caching a few albuns so she can listen to it while working.

even if we could only cache after listen but could separate by albums would be more than enough.

**Comment by heyfolk** [comment]
Why you are not adding a download button to store in local phone storage it will be great if you add download button :)

**Comment by vfsfitvnm** [comment]
Because this isn't a music downloader

**Comment by jerryop** [comment]
This feature is the most important, I got mad, the cache gets deleted every week, I have to listen now again these 100 songs 🥺🥲

**Comment by jerryop** [comment]
> 

This app doesn't have good Ui

**Comment by vfsfitvnm** [comment]
@jerryop you should find a way to whitelist vimusic from that intrusive cleaning operation then. This is definitely not related to a download feature, and it's not a problem I (the programmer) should solve. I can certainly move the cache to another directory, but I won't do that because of a shitty vendor (or whatever) idea

**Comment by ishanveersg** [comment]
@vfsfitvnm I went through the issues, my idea would be to include an option on long pressing a song saying "make available offline" which would continue caching the song even if playback is paused or the user switches to another song. This would solve the problem which I often have which is when I try to head out but want to quickly save a few songs to listen to on a walk or something. I was using the app [Music](https://github.com/z-huang/music) to save songs offline on demand but once i start adding more than 10 songs offline, the apps starts crashing a lot. Your app is mostly flawless in terms of bugs, metadata retrieval and playback speed in my personal experience(thanks a ton for that btw!). I know you are not getting paid for this project so I am trying to find the easiest implementation of an ondemand offline option.

**Comment by vfsfitvnm** [comment]
Caching a single song:

https://user-images.githubusercontent.com/46219656/189527129-2036b2db-f2b9-42e0-87b7-7c80fb0e7f72.mp4


Caching a whole playlist:

https://user-images.githubusercontent.com/46219656/189527091-14cbbb2c-fd79-49f3-bec7-2589e8889fdf.mp4

:thinking:

**Comment by vfsfitvnm** [comment]
However, I don't know if it will ever make to the app. I just don't like the idea.

This feature causes the app make a big step forward being misinterpreted as a music downloader. People will think they need this feature to save a song - you know, this is how any other music app works - and I would hate that. This might be arrogant, but I want to eradicate this common believing.

But I do admit such feature would be helpful in scenarios like https://github.com/vfsfitvnm/ViMusic/issues/83#issuecomment-1217830944. But I don't think it's worth the tradeoff...

**Comment by ikanakova** [comment]
@vfsfitvnm Perhaps the solution would be not present it as a download but only save as a cache. Then people wouldn't see it as a music downloader.

**Comment by jerryop** [comment]
> Caching a single song:
> 
>  device-2022-09-11-141855.mp4 
> Caching a whole playlist:
> 
>  device-2022-09-11-141703.mp4 
> 🤔

This is the best feature to manually select song to make it offline many people's don't want to store all songs as offline they listen in a day most probably we make one or four favourite songs offline to listen whenever we want without any internet I think you should implement this feature:)

**Comment by Lolozweipunktnull** [comment]
I think you should implement that. Exactly such a function is still missing from the app for me personally, but it's already great now, then it would be almost perfect.

Because it is not directly called "download", I don't think anyone misunderstands.

Thank you very much for your work and time for this project.

**Comment by Lolozweipunktnull** [comment]
> However, I don't know if it will ever make to the app. I just don't like the idea.
> 
> This feature causes the app make a big step forward to being misinterpreted as a music downloader. People will think they need this feature to save a song - you know, this is how any other music app works - and I would hate that. This might be arrogant, but I want to eradicate this common believing.
> 
> But I do admit such feature would be helpful in scenarios like [#83 (comment)](https://github.com/vfsfitvnm/ViMusic/issues/83#issuecomment-1217830944). But I don't think it's worth the tradeoff...

I think you should implement that. Exactly that function is still missing from the app for me personally, but it's already great now, then it would be almost perfect.

Because it is not directly called "download", I don't think anyone misunderstands.

Thank you very much for your work and time for this project.

**Comment by DaniAndTheWeb** [comment]
I think that the best way for allowing someone to download conveniently without adding any direct download option would be to add a share button to be able to share the music link to whatever other app the user has to download it.

**Comment by okidoki-jpg** [comment]
> 

Yes! this is what I was looking for . the ability to cache multiple songs or albums at once ahead of time would really be awesome . i assume this feature isn't public yet?

**Comment by okidoki-jpg** [comment]
> Caching a single song:
> 
>  device-2022-09-11-141855.mp4 
> Caching a whole playlist:
> 
>  device-2022-09-11-141703.mp4 
> 🤔

@vfsfitvnm 
Yes! this is what I was looking for . the ability to cache multiple songs or albums at once ahead of time would really be awesome . i assume this feature isn't public yet?

**Comment by vfsfitvnm** [comment]
Nope, I won't do that (https://github.com/vfsfitvnm/ViMusic/issues/74#issuecomment-1173061341)

**Comment by vfsfitvnm** [comment]
> I do get your point though. To combat this, it might make sense to have an indicator of what's cached and what isn't while using the app, instead of just looking at your list with all cached songs.

This is "impossible" to do (https://github.com/vfsfitvnm/ViMusic/issues/60#issuecomment-1181671363). Getting the cache info of each song is a slow operation, that's why cached songs have a dedicated page. Moreover, most users (me included) don't really want to see such info while they are using the app.

**Comment by Jak2k** [comment]
Recently I heared the whole "Liked"-Playlist to make sure, that it's cached for offline use, but sth. on ny phone cleared the cach.

I think android's cache folders aren't the best method to make music available offline. Please add an option to use another folder as cache.

If I remember correct, the cache wasn't in the cache folder in an older version. I don't think it's very complicatet to make a setting for switching between that behaviors.

**Comment by Pentaphon** [comment]
Opus is better quality than AAC so I don't know why people are asking for AAC downloading when Opus is better and only the best Opus stream should be streamed using this app. If the dev ever introduces an Opus download option, it would need to download the Opus stream and repackage it to a music-friendly container like .opus.

**Comment by FaNticDaedlYs** [comment]
Mm I think its confusing for new users to find download option.. most of them don't even know caching....
Also there r bugs in caching as it stop after a moment showing error... so its better to add a offline mode

**Comment by FaNticDaedlYs** [comment]
> Caching a single song:
> 
> https://user-images.githubusercontent.com/46219656/189527129-2036b2db-f2b9-42e0-87b7-7c80fb0e7f72.mp4
> 
> 
> Caching a whole playlist:
> 
> https://user-images.githubusercontent.com/46219656/189527091-14cbbb2c-fd79-49f3-bec7-2589e8889fdf.mp4
> 
> :thinking: 
> 

Ok so if r having issues with adding caching button n stuff... Why don't you just add a simple alg for downloading m4a(most preferred) or MP3 file so that there is music playback and no danger for loss of cached songs...

**Comment by pranshuj73** [comment]
@vfsfitvnm I think a pre-cache feature makes sense, having the offline mode would be a lifesaver for me since I like to listen to music during traveling and studying but I have limited data for the day. If I can just precache my songs I usually loop, that would be more than enough for me. Please consider adding pre-caching to ViMusic.

**Comment by EstebanOD** [comment]
After reading every comment under this issue and under some other issues I think it's clear that this is not meant as a software to primarily use for offline use; it's basically **a frontend for YTMusic** with additional features, but downloading isn't one of those features. It's also clear that _this isn't something the developer will consider adding_.

If you're looking for an app that download songs there's already plenty:

1. [NewPipe](https://github.com/TeamNewPipe/NewPipe) for downloading locally video/audio from YT and choosing codec/bitrate, it has the fastest download speeds too, you can also download from SoundCloud and Bandcamp
2. [LibreTube ](https://github.com/libre-tube/LibreTube) is like NewPipe, but has more modern UI and SponsorBlock integration, and uses [Piped](https://github.com/TeamPiped/Piped)'s API which itself uses NewPipe's [Extractor](https://github.com/TeamNewPipe/NewPipeExtractor), unlike NewPipe the downloads are stored in-app
3.  [InnerTune ](https://github.com/z-huang/InnerTune) to listen and download in-app from YT and YTMusic, works similarily to ViMusic

### Which one to choose?
You want to download songs on your phone and not in-app:
- No = InnerTube/LibreTube (downloaded songs are stored inside the app)
- Yes = NewPipe (downloaded songs are in your files, as .opus or .m4a)
 
You want the app to have a useable YT player like ViMusic:
- Meh = NewPipe
- Yes = InnerTube/LibreTube

Hope this'll help you guys.

**Comment by rancidfrog** [comment]
@vfsfitvnm 

Download can be useful, for example:
I have a song that is in the offline list.
But, the user has made it private.
So now, I cant share it

I do not want to lose it, as it was the only good audio quality of the song.
So, downloading it, would be great.

**Comment by rancidfrog** [comment]
@FaNticDaedlYs 
As my above example explained.
I do not want to lose the cached song as the video was made private.

Also, I cant listen to the song in the current format on other music players.

Having export feature in m4a format would be amazingly useful in a lot of cases

**Comment by rancidfrog** [comment]
@EstebanOD 
But, 
App can still have an export button, which converts cached song to m4a and saves it where user desires

That way it does not interfere with Current app model.
At the same time extends feature which makes all users happy

**Comment by midhun956** [comment]
The cache Feature is nice but the user have very little control. Like having an option to add a specific song to cache without listening to that song would be nice. And another option to convert or download that song to a mp3 or any higher quality file.

**Comment by Radical4ever** [comment]
I rlly think adding this feature could make the app a good upgrade, alongside the possibility to see local music files

**Comment by Inspiredshadow** [comment]
@vfsfitvnm please add an option to support proxy inside the app.

**Comment by ikanakova** [comment]
A played songs indicator to show % cached songs (in the playlist / album) like in Newpipe would be great.

**Comment by BlanXyz** [comment]
> Caching a single song:
> 
> https://user-images.githubusercontent.com/46219656/189527129-2036b2db-f2b9-42e0-87b7-7c80fb0e7f72.mp4
> 
> 
> Caching a whole playlist:
> 
> https://user-images.githubusercontent.com/46219656/189527091-14cbbb2c-fd79-49f3-bec7-2589e8889fdf.mp4
> 
> :thinking: 
> 

The only feature this app is still missing. So close to perfection.. 
Argh, time to learn some Kotlin I guess 🤠

**Comment by amorgado87** [comment]
A button to pre-cache every song/album/playlist will be a great addition. This is the only major feature this app is still missing.

**Comment by HiroDane** [comment]
Since the author doesn't want to add this feature, but it will certainly be useful for majority of the users, can someone else fork a version which adds this and make it a different app?

**Comment by devlocalhost** [comment]
> Since the author doesn't want to add this feature, but it will certainly be useful for majority of the users, can someone else fork a version which adds this and make it a different app?

Local files feature too

**Comment by emkay5771** [comment]
Once a fork is made, wouldn't be super difficult to add other sources (podcasts, full YT, for instance).

**Comment by Pentaphon** [comment]
Doesn't this project seem to be on hold for months now? Not sure if the dev is pushing stuff much less considering this feature.

**Comment by JPGuillemin** [comment]
160kbps OPUS is CD/FLAC quality.
OPUS starts to be undistinguishable from FLAC above 128kbps.

**Comment by AIMBOT312** [comment]
Hey! I'd just like to agree with what others are saying, allowing people to specifically download songs and/or albums for offline playing would be a lovely feature to see implemented. Love the app so far though! :)

**Comment by snakedye** [comment]
![chatgpt](https://user-images.githubusercontent.com/49378990/228598107-f828fecc-9409-416e-8a72-a6bd8351d342.png)

**Comment by spookypie314** [comment]
Yes please add download button as soon as in next update ☺️, like real one because cache files deleted automatically so i can't trust this one,
Your ui is ❤️❤️ just love this, please add download button to store offline and app is is perfect 🙏 [(2)can i share playlist.?]

**Comment by devlocalhost** [comment]
> Yes please add download button as soon as in next update ☺️, like real one because cache files deleted automatically so i can't trust this one, Your ui is ❤️❤️ just love this, please add download button to store offline and app is is perfect 🙏 [(2)can i share playlist.?]

1) there's a way to "download" songs: https://github.com/vfsfitvnm/ViMusic/issues/83#issuecomment-1242952730

2) https://github.com/vfsfitvnm/ViMusic/issues/83#issuecomment-1212209332. Or just increase max cache size

**Comment by devlocalhost** [comment]
> @devlocalhost These are just concept shots, and I am sad that the main dev decided that this is not the way the app should work :/ It's not a thing that makes ViMusic a song downloader.. but yeah.

I understand and I'd like a download button too. (I'd also like a local files feature). But hey, at least there's a workaround to "somewhat download" a song.

If the main dev doesn't want this feature or hasn't told us anything about it, then I'd say we need to stop "begging" for this feature

**Comment by devlocalhost** [comment]
> There are some forks of the app that probably have download feature. I'd like to manage the cache too, but since vfsfitvnm has switched the app state to maintenance only, we can't expect anything cache-related to be implemented in the near future. If ever.. :D

I looked at this repos fork once, and I didn't found anything. They were just forks, no modifications

**Comment by spookypie314** [comment]
> Caching a single song:
> 
> Caching a whole playlist:
> 
> 🤔

Ok this is great feature, but why I'm not getting this option.? My app is up to date with 0.5.4
And how can i prevent these cache files from deleting automatically
Thanks🙏
![Screenshot_20230404_064233](https://user-images.githubusercontent.com/124502842/229660728-7e85032a-2f96-462c-ac78-2cffde2b77b8.jpg)
![Screenshot_20230404_064228](https://user-images.githubusercontent.com/124502842/229660734-5f562690-807a-4e26-b5b4-df7d12f3ae67.jpg)

**Comment by Henry-Hiles** [comment]
Its such a shame this didnt exist, if it did then ViMusic would be perfect. This is the only thing that keeps my family using Spotify instead.

**Comment by Jak2k** [comment]
The last commit was in December 22. As I can see, the app isn't mantained anymore. Maybe someone should fork it.

**Comment by spookypie314** [comment]
> > Caching a single song:
> > 
> > 
> >   
> > Caching a whole playlist:
> > 
> > 
> >   
> > 🤔
> 
> The only feature this app is still missing. So close to perfection.. Argh, time to learn some Kotlin I guess 🤠

Which app is this.?
Is this vimusic.?

**Comment by spookypie314** [comment]
> @spookypie314 These shots were just a concept by @vfsfitvnm. The feature was never implemented.

Oh, ro it seems like developer doesn't want to add this feature, good thing about this app was ui, but i have to check other app if there( used innertune it's good but sometimes it not stable)🙏

**Comment by Melvinfused** [comment]
> Because this isn't a music downloader

By downloading I think he meant to be able to cache the album or playlist to play offline in the app itself (I would appreciate something like that). I don't ask for downloading M4A or other files.

**Comment by Alphamars** [comment]
I had a difficult experience using ViMusic,  I played all the songs just to be able to have them as cache and it was working fine until I accidentally clear cache when phone cleaner automatically request for cleaning. Since then wasn't able to play those songs.. Is there a way to save only what you want instead? Most time I start radio and go to work whenever I hear something I like I click like but the rest would be save to cache and that is like another huge work to start deleting. Please consider a download button and everything we played completely is not saved on cache.

**Comment by lazydevyo** [comment]
Add a download button for playlists and songs like Spotify has. I want to listen on the go, I don't want to sit and listen to a song at home, wait till it finishes playing so I can actually listen to it on the road.
